File: 61-rotate-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        // Handle empty list or single node
        if (!head || !head->next || k == 0) {
            return head;
        }

        // 1. Compute the length of the list and find the tail
        ListNode* tail = head;
        int n = 1;
        while (tail->next) {
            tail = tail->next;
            n++;
        }

        // 2. Calculate effective rotation
        k = k % n;
        if (k == 0) return head;

        // 3. Make the list circular
        tail->next = head;

        // 4. Find the new tail: it is at (n - k - 1) steps from the current head
        // Or (n - k) steps from the current tail
        int stepsToNewTail = n - k;
        ListNode* newTail = tail;
        while (stepsToNewTail--) {
            newTail = newTail->next;
        }

        // 5. Break the circle and set the new head
        ListNode* newHead = newTail->next;
        newTail->next = nullptr;

        return newHead;
    }
};

File: 61-rotate-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # Handle edge cases: empty list, single node, or no rotation needed
        if not head or not head.next or k == 0:
            return head
        
        # 1. Determine the length of the list and locate the tail
        tail = head
        length = 1
        while tail.next:
            tail = tail.next
            length += 1
            
        # 2. Calculate the effective number of rotations
        k = k % length
        if k == 0:
            return head
            
        # 3. Connect the tail to the head to form a cycle
        tail.next = head
        
        # 4. Find the node that will become the new tail
        # The new head is at index (length - k), so the new tail is at (length - k - 1)
        # Moving (length - k) steps from the current tail reaches the new tail
        steps_to_new_tail = length - k
        new_tail = tail
        for _ in range(steps_to_new_tail):
            new_tail = new_tail.next
            
        # 5. Establish the new head and break the circular connection
        new_head = new_tail.next
        new_tail.next = None
        
        return new_head